common/src/scoop/mod.rs

```diff
@@ -0,0 +1,89 @@
+mod source;
+
+pub use source::*;
+
+use crate::progress::{Progress, ProgressBar};
+use anyhow::bail;
+use bytes::Bytes;
+use futures_util::{StreamExt, TryStreamExt, stream};
+use tracing::instrument;
+
+/// A tool to build a [`Scooper`].
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct ScooperBuilder {
+    pub sources: Vec<Source>,
+    pub delete: bool,
+    pub processed: Option<String>,
+    pub failed: Option<String>,
+}
+
+impl ScooperBuilder {
+    pub async fn build(self) -> anyhow::Result<Scooper> {
+        let sources = self.discover().await?;
+        Ok(Scooper {
+            builder: self,
+            sources,
+        })
+    }
+
+    /// Discover files to upload
+    async fn discover(&self) -> anyhow::Result<Vec<Source>> {
+        Ok(stream::iter(&self.sources)
+            .then(async |source| source.clone().discover().await)
+            .try_collect::<Vec<_>>()
+            .await?
+            .into_iter()
+            .flatten()
+            .collect())
+    }
+}
+
+/// A tool to scoop up files
+pub struct Scooper {
+    builder: ScooperBuilder,
+    sources: Vec<Source>,
+}
+
+impl Scooper {
+    #[instrument(skip_all, err)]
+    pub async fn process<F, P>(self, progress: P, processor: F) -> anyhow::Result<()>
+    where
+        for<'a> F: AsyncFn(&'a str, Bytes) -> anyhow::Result<()> + 'a,
+        P: Progress,
+    {
+        let total = self.sources.len();
+        let mut errors = 0usize;
```

Comment:
what is 0usize ?

Reply:
zero as `usize`. Like `0f32` (zero as `f32`).

Would be the same as:

```rust
let mut errors : usize = 0;
```